Auxiliary Node Gadgets

mattigruener opened the pull request:
Hey John,

just wanted to put this up to see if this is roughly where you wanted to see this going. There's some \todo tags in the code that I need to address and the circular node frames are -ahem- rectangular at the moment. I wasn't sure if you would want to tesselate the circle into triangles or rather smoothstep the radius of the circle in a shader? So still stuff to do, but I thought it's worth chatting about this before doing any more work on it.

Cheerio,

Matti.

johnhaddon commented on include/Gaffer/AuxiliaryNode.h:
```diff
@@ -0,0 +1,65 @@
+
+// AuxiliaryNodes are nodes that only compute input values for other nodes.
+// Animation curves or Expressions would be prominent examples.
+class GAFFER_API AuxiliaryNode : public ComputeNode
```
Hmm, I wasn't expecting any change to the nodes themselves. I thought we were just going to register some metadata for a different sort of gadget for the nodes that needed them. My main concern would be that it could make it harder to implement some other node types in the future - "I need to derive from X, but I can't because it doesn't derive from AuxiliaryNode". Basically, this breaks the separation between the UI and the non-UI code, by shoehorning a class for the UI right into the middle of the class hierarchy...

johnhaddon commented on include/GafferUI/AuxiliaryNodeGadget.h:
```diff
@@ -0,0 +1,81 @@
+//////////////////////////////////////////////////////////////////////////
+//
+//  Copyright (c) 2017, Image Engine Design Inc. All rights reserved.
```
2018

johnhaddon commented on include/GafferUI/AuxiliaryNodeGadget.h:
```diff
@@ -0,0 +1,81 @@
+namespace GafferUI
+{
+
+class GAFFERUI_API AuxiliaryNodeGadget : public StandardNodeGadget
```
Any reason we're deriving from StandardNodeGadget? Since we will never have nodules, I was thinking we might be able to just derive from NodeGadget instead, and end up with a more lightweight gadget without lots of internal gadget layout going unused.

johnhaddon commented on include/GafferUI/AuxiliaryNodeGadget.h:
```diff
@@ -0,0 +1,81 @@
+
+		IE_CORE_DECLARERUNTIMETYPEDEXTENSION( GafferUI::AuxiliaryNodeGadget, AuxiliaryNodeGadgetTypeId, StandardNodeGadget );
+
+		AuxiliaryNodeGadget( Gaffer::NodePtr node );
```
I don't think this is relevant given my comments about the AuxiliaryNode, but presumably the intention of AuxiliaryNodeGadget was to only work with for AuxiliaryNodes. In that case it would be better to have the constructor enforce that constraint via the signature than to do internal checks.

Perhaps this is a problem with the NodeGadgetTypeDescription registration mechanism though, and it's requiring a constructor taking NodePtr? That could be fixed I think, if needs be. But like i said, I don't think we want AuxiliaryNode to exist anyway.

johnhaddon commented on src/GafferUI/AuxiliaryNodeGadget.cpp:
```diff
@@ -0,0 +1,104 @@
+	:	StandardNodeGadget( node )
+{
+	setContents( nullptr );  // remove NameGadget
+	if( !dynamic_cast<Gaffer::AuxiliaryNode *>( node.get() ) )
```
Again, not relevant given previous comments, but this should have been a `runTimeCast()`.

johnhaddon commented on python/GafferUI/AnimationUI.py:
```diff
@@ -50,7 +52,9 @@
 	""",
 
-	"nodeGadget:type", "",
+	"nodeGadget:type", "GafferUI::AuxiliaryNodeGadget",
```
This is exactly what I was expecting to see - use "nodeGadget:type" to register the kind of UI we want. No need for AuxiliaryNode whatsoever if we're doing this anyway.

johnhaddon commented on python/GafferUI/RandomUI.py:
```diff
@@ -67,6 +73,7 @@
 			values are different.
 			""",
 
+			"nodule:type", "",
```
Rather than do this for every plug, we'd probably be better to use a [wildcarded registration](https://github.com/GafferHQ/gaffer/blob/master/python/GafferSceneUI/SceneNodeUI.py#L55). But actually, if we don't derive AuxiliaryNodeGadget from StandardNodeGadget we don't have this issue anyway, because it won't have any mechanism for managing nodules at all.

johnhaddon commented on src/GafferUI/StandardStyle.cpp:
```diff
@@ -957,6 +974,12 @@ void StandardStyle::renderLine( const IECore::LineSegment3f &line ) const
 
+void StandardStyle::renderSolidCircle( float radius ) const
+{
+	// \todo: make this a circle.
```
Take a look at how the DotNodeGadget tricks the system into giving it a circle, or how [BoxIO does something similar](https://github.com/GafferHQ/gaffer/blob/master/python/GafferUI/BoxIOUI.py#L54).

johnhaddon commented on include/GafferUI/StandardStyle.h:
```diff
@@ -91,6 +92,7 @@ class GAFFERUI_API StandardStyle : public Style
 		void renderConnection( const Imath::V3f &srcPosition, const Imath::V3f &srcTangent, const Imath::V3f &dstPosition, const Imath::V3f &dstTangent, State state = NormalState, const Imath::Color3f *userColor = nullptr ) const override;
 		Imath::V3f closestPointOnConnection( const Imath::V3f &p, const Imath::V3f &srcPosition, const Imath::V3f &srcTangent, const Imath::V3f &dstPosition, const Imath::V3f &dstTangent ) const override;
 		void renderAuxiliaryConnection( const Imath::Box2f &srcNodeFrame, const Imath::Box2f &dstNodeFrame ) const override;
+		void renderAuxiliaryNode( float radius, char label, State state, const Imath::Color3f *userColor ) const override;
```
I was kinda thinking of taking a shortcut here, and using the same trick that DotNodeGadget uses to get the circles. Not against this though if Andrew's happy to bump major version soon : it definitely gives more scope to the Style to style things. On the question of scope, perhaps a `const Box2f &bound` is less prescriptive, allowing the style to use circles if it wants, but also rectangles or something else if not? And perhaps the labelling should be a separate thing from the frame, as it is for StandardNodeGadget and `renderNodeFrame()`? i.e. We could just use `renderText()` from within `AuxiliaryNodeGadget::doRenderLayer()`?

mattigruener commented:
> I know the timezones pose a problem and I'm not always as available as would be ideal, so perhaps we need to standardise on having some basic written notes agreed before work starts?

Ok, let's try that.

I think I've addressed all of your comments. 

 - How do you feel about hard-coding the color like this? Deriving from NodeGadget directly removes the mechanism to read in an `m_userColor` and by using the shortcut to draw a circle, I'm kinda taking the StandardStyle out of the equation a little. Are you ok with this or would you rather have that color stored on the style somehow?

- I've tried to replicate the metadata update stuff in StandardNodeGadget to avoid superfluous render calls. I think that makes sense?

Thanks,

Matti.

johnhaddon commented:
> I think I've addressed all of your comments.

Thanks Matti! I am liking the simplicity of this approach, especially when I look again at how much baggage StandardNodeGadget has.

> How do you feel about hard-coding the color like this? Deriving from NodeGadget directly removes the mechanism to read in an m_userColor and by using the shortcut to draw a circle, I'm kinda taking the StandardStyle out of the equation a little.

I think it would be nice to bring back the support for the user color by replicating that small bit of metadata logic out of StandardNodeGadget. There might be an argument for pushing the metadata colour handling onto the NodeGadget base class, but let's just duplicate for now and see how it pans out.

> Are you ok with this or would you rather have that color stored on the style somehow?

In practice the node colours aren't really provided by the style anyway, since they're always specified by the metadata, so I think this is fine. If you wanted to reintroduce the `Style::renderAuxiliaryNodeFrame()` method so the style at least has some leeway, then that'd be fine (perhaps with the alternate signature I suggested).

> I've tried to replicate the metadata update stuff in StandardNodeGadget to avoid superfluous render calls. I think that makes sense?

Yep, definitely.

Cheers...
John

mattigruener commented:
Hey John,

I've added the mechanism for reading the color from metadata. The UI files have been changed accordingly - I'll send you a screenshot so you don't have to compile this code just to see what it looks like. Can't paste the image into the comment here for security reasons...

Thanks :)

Matti.

johnhaddon commented:
Thanks Matti!
